File: aspnet-in-practice/08-09/CS/CoolMVBlog/Areas/Backoffice/Controllers/PostController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.Mvc;
using CoolMvcBlog.Models;
using CoolMvcBlog.Areas.Backoffice.Models;
using System.Data;
using System.Data.Objects;
using CoolMvcBlog.Utils;

namespace CoolMvcBlog.Areas.Backoffice.Controllers
{
    public class PostController : Controller
    {
        public ActionResult Index()
        {
            return this.View(ObjectContextModule.CurrentContext
                    .PostSet
                    .Include("Author")
                    .OrderByDescending(p => p.DatePublished)
                    .ToList());
        }

        [HttpGet]
        public ActionResult Edit(int id)
        {
            this.ViewData["Authors"] = AuthorsService.GetAuthors();
            this.ViewData["Categories"] = CategoriesService.GetCategories();
            return View(ObjectContextModule.CurrentContext
                .PostSet
                .Include("Author")
                .Include("Categories")
                .Where(p => p.Id == id)
                .Single());
        }

        [HttpPost]
        public ActionResult Edit(Post post)
        {
            if (this.ModelState.IsValid)
            {
                ObjectContextModule.CurrentContext.SaveChanges();
                return this.RedirectToAction("Index");
            }

            this.ViewData["Authors"] = AuthorsService.GetAuthors();
            this.ViewData["Categories"] = CategoriesService.GetCategories();

            return this.View(post);
        }

        //[HttpPost]
        //public ActionResult Edit(Post post)
        //{
        //    if (this.ModelState.IsValid)
        //    {
        //        using (BlogModelContainer ctx = new BlogModelContainer())
        //        {
        //            var original = ctx.PostSet
        //                .Where(p => p.Id == post.Id)
        //                .Single();

        //            if (this.TryUpdateModel(original))
        //            {
        //                ctx.SaveChanges();
        //                return this.RedirectToAction("Index");
        //            }
        //        }
        //    }

        //    this.ViewData["Authors"] = AuthorsService.GetAuthors();
        //    this.ViewData["Categories"] = CategoriesService.GetCategories();

        //    return this.View(post);
        //}
    }
}

File: aspnet-in-practice/08-09/VB/CoolMVCBlog/Areas/Backoffice/Controllers/PostController.vb
Namespace CoolMvcBlog.Areas.Backoffice.Controllers
    Public Class PostController
        Inherits Controller

        Public Function Index() As ActionResult
            Return Me.View(ObjectContextModule.CurrentContext.
                           PostSet.
                           Include("Author").
                           OrderByDescending(Function(p) p.DatePublished).
                           ToList)
        End Function

        <HttpGet()>
        Public Function Edit(ByVal id As Integer) As ActionResult
            Me.ViewData("Authors") = AuthorsService.GetAuthors()
            Me.ViewData("Categories") = CategoriesService.GetCategories()

            Return View(ObjectContextModule.CurrentContext.
                        PostSet.
                        Include("Author").
                        Include("Categories").
                        Where(Function(p) p.Id = id).
                        Single())
        End Function

        <HttpPost()>
        Public Function Edit(ByVal post As Post) As ActionResult
            If Me.ModelState.IsValid Then
                ObjectContextModule.CurrentContext.SaveChanges()
                Return Me.RedirectToAction("Index")
            End If

            Me.ViewData("Authors") = AuthorsService.GetAuthors()
            Me.ViewData("Categories") = CategoriesService.GetCategories()

            Return Me.View(post)
        End Function

        '<HttpPost()>
        'Public Function Edit(ByVal post As Post) As ActionResult
        '    If Me.ModelState.IsValid Then
        '        Using ctx As New BlogModelContainer
        '            Dim original = ctx.PostSet.
        '                Where(Function(p) p.Id = post.Id).
        '                Single

        '            If Me.TryUpdateModel(original) Then
        '                ctx.SaveChanges()
        '                Return Me.RedirectToAction("Index")
        '            End If
        '        End Using
        '    End If

        '    Me.ViewData("Authors") = AuthorsService.GetAuthors()
        '    Me.ViewData("Categories") = CategoriesService.GetCategories()

        '    Return Me.View(post)
        'End Function

    End Class
End Namespace